Fixed value-rounding for @2x+ devices

ebelair opened the pull request:
## Problem
The methods `ceilFloatIfRequired` and `floorFloatIfRequired` were only aligning coordinates to the display grid for **@1x** devices, by rounding values to whole pixels.

For **@2x** and **@3x** devices, the views still need to be aligned to the grid, otherwise they may look blurry. The alignment is made to **half** and **third** of pixels, respectively (`0.5` and `0.3333` steps).

## Solution
I changed these two methods to round the values intelligently, and renamed them accordingly (`ceilFloatToDisplayScale` and `floorFloatToDisplayScale`).

I just wanted to validate the strategy used: multiplying by the ratio before rounding and dividing afterwards might be an expensive operation, though I couldn't find a more appropriate solution.

> **Note:** I still kept the `onNonRetina` validation to prevent doing `round(x * 1.0f) / 1.0f` uselessly, especially since non-retina devices *should technically* be the least performant.

marclefrancois commented:
:+1:

flambert commented:
You could compute the "inverseDisplayScale" and multiply by it instead of dividing by "displayScale" to shave off more CPU cycles. :)

flambert commented:
You could implement the "+ (void)load" method that is called once when initializing the app, compute the "displayScale" and "inverseDisplayScale" once there (instead of using dispatch_once).

In this method, you could also keep a pointer to the correct rounding functions and call them instead of the "ceilFloatToDisplayScale" and "floorFloatToDisplayScale" methods.

For example, on retina, the rounding function pointer would point to a function that does "roundf(value * scale) * inverseScale" and on non-retina it would point to "roundf" directly.

So you save the call to the method, the dispatch_once call and the onNonRetina test.

ebelair commented:
All true. But can we change the `+ (void)load` method of an app *inside* a pod?

Or could be simply use macros instead?

```objc
#define  DISPLAY_SCALE      [UIScreen mainScreen].scale
#define  DISPLAY_SCALE_INV  1.0f / DISPLAY_SCALE

#define  CEIL_SCALE(x)      ceilf(x * DISPLAY_SCALE) * DISPLAY_SCALE_INV
#define  FLOOR_SCALE(x)     floorf(x * DISPLAY_SCALE) * DISPLAY_SCALE_INV
```

flambert commented:
```+ (void)load``` is called for each categories, I don't see why it would make a difference because the category is packaged in a pod, a framework or directly in the app.

Macros are replaced at compile time, it will make it worse since you will continually get the scale, inverse it and then multiply it.

Each time you use it like that:

```
x = CEIL_SCALE(x);
```

You get that after the pre-processor is done:

```
x = ceilf(x * [UIScreen mainScreen].scale) * 1.0f / [UIScreen mainScreen].scale;
```

ebelair commented:
Alright I didn't know the class-specific `+ (void)load` method would just be called once, I learned something.

So how about this:

```objc
@implementation MCUIViewLayoutPosition

CGFloat displayScale;
CGFloat displayScaleInv;
CGFloat (^ceilFloatToDisplayScale)(CGFloat x);
CGFloat (^floorFloatToDisplayScale)(CGFloat x);

+ (void)load {
    displayScale = [UIScreen mainScreen].scale;
    displayScaleInv = 1.0f / displayScale;
    
    if (displayScale > 1.0f) {
        ceilFloatToDisplayScale = ^(CGFloat x) {
            return ceilf(x * displayScale) * displayScaleInv;
        };
        floorFloatToDisplayScale = ^(CGFloat x) {
            return floorf(x * displayScale) * displayScaleInv;
        };
    } else {
        ceilFloatToDisplayScale = ^(CGFloat x) {
            return (CGFloat)ceilf(x);
        };
        floorFloatToDisplayScale = ^(CGFloat x) {
            return (CGFloat)floorf(x);
        };
    }
}

[...]
```

ebelair commented:
Alright I didn't know the class-specific `+ (void)load` method would just be called once, I learned something.

I moved the implementations to static block functions and added public methods to be able to use them in projects, as discussed verbally.
